### PR TITLE
Add backup extension implementation for AWS provider

### PR DESCRIPTION
**Pull request by swapnilgm**
**What this PR does / why we need it**:
This PR add the backup extension resources to Gardener as per (backup infrastructure proposal)[https://github.com/gardener/gardener/blob/master/docs/proposals/02-backupinfra.md]
Ref: 
1. gardener/etcd-backup-restore#141
2. https://github.com/gardener/gardener/pull/1079

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Garden-extension now supports the backup-extension resource for AWS provider.
```

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -9,7 +9,8 @@ required = [
 
 [[override]]
   name = "github.com/gardener/gardener"
-  revision = "8725569dc996defe240950cc71cc4509a29c91fe"
+  source = "github.com/swapnilgm/gardener"
```
Comment (rfranzke) [inline Gopkg.toml]: Once https://github.com/gardener/gardener/pull/1079 is merged you should file another PR for the https://github.com/gardener/gardener/tree/gardener-extensions-fixes branch

File: controllers/provider-aws/charts/provider-aws/doc.go [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:generate ../../../../hack/generate-controller-registration.sh provider-aws . ../../example/controller-registration.yaml Infrastructure:aws ControlPlane:aws Worker:aws
+//go:generate ../../../../hack/generate-controller-registration.sh provider-aws . ../../example/controller-registration.yaml  BackupBucket:aws BackupEntry:aws ControlPlane:aws Infrastructure:aws Worker:aws
```
Comment (rfranzke) [inline controllers/provider-aws/charts/provider-aws/doc.go]:
```suggestion
//go:generate ../../../../hack/generate-controller-registration.sh provider-aws . ../../example/controller-registration.yaml BackupBucket:aws BackupEntry:aws ControlPlane:aws Infrastructure:aws Worker:aws
```

File: controllers/provider-aws/charts/provider-aws/values.yaml [hunk 3]
```diff
@@ -6,6 +6,11 @@ image:
 resources: {}
 
 controllers:
+  backupbucket:
+    concurrentSyncs: 5
+  backupentry:
+    concurrentSyncs: 5
+    deletionGracePeriodDays: false
```
Comment (rfranzke) [inline controllers/provider-aws/charts/provider-aws/values.yaml]: true/false? Shouldn't we have an `int` here?

File: controllers/provider-aws/cmd/gardener-extension-provider-aws/app/app.go [hunk 4]
```diff
@@ -66,6 +83,11 @@ func NewControllerManagerCommand(ctx context.Context) *cobra.Command {
 			MaxConcurrentReconciles: 5,
 		}
 
+		// // options for the controlplane controller
+		// backupBucketCtrlOpts = &controllercmd.ControllerOptions{
```
Comment (rfranzke) [inline controllers/provider-aws/cmd/gardener-extension-provider-aws/app/app.go]: Why is this here?

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	if err := a.client.Get(ctx, kutil.Key(bb.Spec.SecretRef.Namespace, bb.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), bb.Spec.Region)
+	if err != nil {
+		return err
+	}
+
+	return awsClient.DeleteBucket(ctx, bb.Name)
```
Comment (rfranzke) [inline controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: Before deleting the bucket, shouldn't you check that no `BackupEntry` objects reference this bucket anymore?

File: controllers/provider-aws/pkg/controller/backupentry/actuator.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	return &actuator{}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+func (a *actuator) Reconcile(ctx context.Context, be *extensionsv1alpha1.BackupEntry) error {
+	return nil
```
Comment (rfranzke) [inline controllers/provider-aws/pkg/controller/backupentry/actuator.go]: to-be-implemented, right?

File: pkg/controller/backupbucket/controller.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	// Predicates are the predicates to use.
+	// If unset, GenerationChangedPredicate will be used.
+	Predicates []predicate.Predicate
+}
+
+// DefaultPredicates returns the default predicates for a controlplane reconciler.
+func DefaultPredicates(mgr manager.Manager) []predicate.Predicate {
+	return []predicate.Predicate{
+		extensionscontroller.GenerationChangedPredicate(),
+		//TODO: Do we need force deletion? extensionscontroller.AnnotationsChangedPredicate(),
```
Comment (rfranzke) [inline pkg/controller/backupbucket/controller.go]: Hm, good question. Although we might not often use it it would be a nice feature to have, right?!

File: pkg/controller/backupbucket/reconciler.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		r.logger.Error(err, "Error removing finalizer from backupbucket", "backupbucket", bb.Name)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) updateStatusProcessing(ctx context.Context, bb *extensionsv1alpha1.BackupBucket, lastOperationType gardencorev1alpha1.LastOperationType, description string) error {
+	bb.Status.LastOperation = extensionscontroller.LastOperation(lastOperationType, gardencorev1alpha1.LastOperationStateProcessing, 1, description)
+	return r.client.Status().Update(ctx, bb)
```
Comment (rfranzke) [inline pkg/controller/backupbucket/reconciler.go]: Use the `TryUpdateStatus` functions like in the other controllers.

File: pkg/controller/backupentry/controller.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	// Predicates are the predicates to use.
+	// If unset, GenerationChangedPredicate will be used.
+	Predicates []predicate.Predicate
+}
+
+// DefaultPredicates returns the default predicates for a controlplane reconciler.
+func DefaultPredicates(mgr manager.Manager) []predicate.Predicate {
+	return []predicate.Predicate{
+		extensionscontroller.GenerationChangedPredicate(),
+		//TODO: Do we need force deletion? extensionscontroller.AnnotationsChangedPredicate(),
```
Comment (rfranzke) [inline pkg/controller/backupentry/controller.go]: See above: would be nice to have

File: controllers/provider-aws/charts/provider-aws/values.yaml (hunk 3 above)
Comment (swapnilgm) [reply controllers/provider-aws/charts/provider-aws/values.yaml]: :+1:

File: Gopkg.toml (hunk 1 above)
Comment (swapnilgm) [reply Gopkg.toml]: Yeah Sure.

File: controllers/provider-aws/cmd/gardener-extension-provider-aws/app/app.go (hunk 4 above)
Comment (swapnilgm) [reply controllers/provider-aws/cmd/gardener-extension-provider-aws/app/app.go]: I guess it came from old commit while rebase. I'll clear that. Strange it didn't give any error in `make verify`

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go (hunk 5 above)
Comment (swapnilgm) [reply controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: I plan to have that check only for BackupBucket resource in Garden cluster i.e. as a part of BackupBucket controller of GCM and not gardener-extension.

Comment (rfranzke) [reply controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: Hm, not sure if that's a good idea to only do it that way. IMHO you need both checks to be safer.

File: controllers/provider-aws/pkg/controller/backupentry/actuator.go (hunk 6 above)
Comment (swapnilgm) [reply controllers/provider-aws/pkg/controller/backupentry/actuator.go]: Yes. It would be mostly related to backword compatibility.

File: pkg/controller/backupbucket/controller.go (hunk 7 above)
Comment (swapnilgm) [reply pkg/controller/backupbucket/controller.go]: Cool. I'll implement that.

File: pkg/controller/backupbucket/reconciler.go (hunk 8 above)
Comment (swapnilgm) [reply pkg/controller/backupbucket/reconciler.go]: I referred (this)[https://github.com/gardener/gardener-extensions/blob/master/pkg/controller/controlplane/reconciler.go#L169] Anyways, if that's the right way then i'll change it at both places.

Comment (rfranzke) [reply pkg/controller/backupbucket/reconciler.go]: Hm, ok, it should be changed in the control plane controller as well (cc @stoyanr). Let's do that separately so that you can focus on the backup stuff.

File: pkg/controller/backupbucket/controller.go (hunk 7 above)
Comment (swapnilgm) [reply pkg/controller/backupbucket/controller.go]: It's no more needed. Since deletionGracePeriod logic is completely moved to GCM. The extension/backupEntry resource is marked for deletion by gcm-core/backupEntry controller only after deletionGracePeriod is over. So, the annotation support is on gcm-core/backupEntry rather thant extension/backupEntry Its deletionGracePeriod ism better kept generic rather than extension specific.

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go (hunk 5 above)
Comment (swapnilgm) [reply controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: I haven't added it here. Cause backupEntry might be in different seed other than backup bucket. And hence backup bucket won't have access to all backupEntries. The check in GCM for the same handles that safely.

**Comment by swapnilgm** [comment]
@rfranzke Thank you for your inputs. The PR is ready for review, except the vendoring part, which will be fixed post merging of PR https://github.com/gardener/gardener/pull/1122.

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+	return awsClient.CreateBucket(ctx, bb.Name)
+}
+
+func (a *actuator) Delete(ctx context.Context, bb *extensionsv1alpha1.BackupBucket) error {
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(bb.Spec.SecretRef.Namespace, bb.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), bb.Spec.Region)
```
Comment (rfranzke) [inline controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: Maybe move the reading of the provider secret + creation of an AWS client into a separate private function that takes the secret references and returns the AWS client?

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go (hunk 5 above)
Comment (rfranzke) [reply controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: Thanks, makes sense, agreed.

File: controllers/provider-aws/pkg/controller/backupentry/actuator.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+	return backupSecretData, nil
+}
+
+func (a *actuator) Delete(ctx context.Context, be *extensionsv1alpha1.BackupEntry) error {
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(be.Spec.SecretRef.Namespace, be.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), be.Spec.Region)
```
Comment (rfranzke) [inline controllers/provider-aws/pkg/controller/backupentry/actuator.go]: See above, the utility function would help you to write the code only once which you now wrote three times.

File: pkg/controller/backupbucket/controller.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, options controller.Options, predicates []predicate.Predicate) error {
+	ctrl, err := controller.New(ControllerName, mgr, options)
+	if err != nil {
+		return err
+	}
+
+	return ctrl.Watch(&source.Kind{Type: &extensionsv1alpha1.BackupBucket{}}, &handler.EnqueueRequestForObject{}, predicates...)
```
Comment (rfranzke) [inline pkg/controller/backupbucket/controller.go]: Please also establish a watch for the referenced secrets (like in the other controllers).

File: pkg/controller/backupentry/actuator.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+// ProviderActuator preforms provider specific operation with BackupBucket resources.
+type ProviderActuator interface {
+	// Delete deletes the BackupBucket.
+	Delete(context.Context, *extensionsv1alpha1.BackupEntry) error
+	// GetETCDSecretData return the updated the secret data as per provider requirement.
+	GetETCDSecretData(context.Context, *extensionsv1alpha1.BackupEntry, map[string][]byte) (map[string][]byte, error)
+}
+
+type actuator struct {
```
Comment (rfranzke) [inline pkg/controller/backupentry/actuator.go]: Similar to `pkg/worker/genericactuator` or `pkg/controlplane/genericactuator` can you please move the `providerActuator` into a separate package? It might also makes sense thinking about renaming it to `genericactuator`.

File: pkg/controller/backupentry/actuator.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	Reconcile(context.Context, *extensionsv1alpha1.BackupEntry) error
+	// Delete deletes the BackupEntry.
+	Delete(context.Context, *extensionsv1alpha1.BackupEntry) error
+}
+
+// ProviderActuator preforms provider specific operation with BackupBucket resources.
+type ProviderActuator interface {
+	// Delete deletes the BackupBucket.
+	Delete(context.Context, *extensionsv1alpha1.BackupEntry) error
+	// GetETCDSecretData return the updated the secret data as per provider requirement.
```
Comment (rfranzke) [inline pkg/controller/backupentry/actuator.go]:
```suggestion
	// GetETCDSecretData returns the updated secret data as per provider requirement.
```

File: pkg/controller/backupentry/actuator.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	return nil
+}
+
+// InjectFunc enables injecting Kubernetes dependencies into actuator's dependencies.
+func (a *actuator) InjectFunc(f inject.Func) error {
+	return f(a.providerActuator)
+}
+
+// NewActuator creates a new Actuator that updates the status of the handled BackupBucket resources.
+func NewActuator(provideActuator ProviderActuator, logger logr.Logger) Actuator {
```
Comment (rfranzke) [inline pkg/controller/backupentry/actuator.go]:
```suggestion
func NewActuator(providerActuator ProviderActuator, logger logr.Logger) Actuator {
```

File: pkg/controller/backupentry/actuator.go [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+func (a *actuator) Delete(ctx context.Context, be *extensionsv1alpha1.BackupEntry) error {
+	return a.providerActuator.Delete(ctx, be)
+}
+
+// extractShootDetailsFromBackupEntryName returns Shoot resource technicalID its UID from provided <backupEntryName>.
+func extractShootDetailsFromBackupEntryName(backupEntryName string) (shootTechnicalID, shootUID string) {
+	tokens := strings.Split(backupEntryName, "--")
+	shootUID = tokens[len(tokens)-1]
+	shootTechnicalID = strings.TrimSuffix(backupEntryName, shootUID)
+	shootTechnicalID = strings.TrimSuffix(shootTechnicalID, "--")
```
Comment (rfranzke) [inline pkg/controller/backupentry/actuator.go]: Why do you call `TrimSuffix` twice?

```suggestion
	shootTechnicalID = strings.TrimSuffix(shootTechnicalID, "--" + shootUID)
```

File: pkg/controller/backupentry/actuator.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+// Reconcile reconciles the update of a BackupEntry
+func (a *actuator) Reconcile(ctx context.Context, be *extensionsv1alpha1.BackupEntry) error {
+	return a.deployEtcdBackupSecret(ctx, be)
+}
+
+func (a *actuator) deployEtcdBackupSecret(ctx context.Context, be *extensionsv1alpha1.BackupEntry) error {
+	shootTechnicalID, _ := extractShootDetailsFromBackupEntryName(be.Name)
+	namespace := &corev1.Namespace{}
+	if err := a.client.Get(ctx, kutil.Key(shootTechnicalID), namespace); err != nil {
+		if apierrors.IsNotFound(err) {
```
Comment (rfranzke) [inline pkg/controller/backupentry/actuator.go]: Is this a case we really need to handle this way? Can you explain why we should not return an error in this situation?

File: pkg/controller/backupentry/mapper.go [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+	return requests
+}
+
+// SecretToBackupEntryMapper returns a mapper that returns requests for BackupEntry whose
+// referenced secrets have been modified.
+func SecretToBackupEntryMapper(client client.Client, predicates []predicate.Predicate) handler.Mapper {
+	return &secretToBackupEntryMapper{client, predicates}
+}
+
+type namespaceToBackupEntryMapper struct {
```
Comment (rfranzke) [inline pkg/controller/backupentry/mapper.go]: Why do we need this namespace mapper?

File: pkg/controller/backupbucket/controller.go (hunk 12 above)
Comment (swapnilgm) [reply pkg/controller/backupbucket/controller.go]: hmm. I just added for backupEntry.  Will add it.

File: pkg/controller/backupentry/actuator.go (hunk 13 above)
Comment (swapnilgm) [reply pkg/controller/backupentry/actuator.go]: Sure.

File: controllers/provider-aws/pkg/aws/client/client.go [hunk 19]
```diff
@@ -172,3 +174,72 @@ func (c *Client) DeleteSecurityGroup(ctx context.Context, id string) error {
...
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+// CreateBucket creates the s3 bucket with name <bucket>. If it already exist,
+// no error is returned.
+func (c *Client) CreateBucket(ctx context.Context, bucket string) error {
```
Comment (adracus) [inline controllers/provider-aws/pkg/aws/client/client.go]: I'd recommend renaming this to `CreateBucketIfNotExists`

File: controllers/provider-aws/pkg/aws/client/client.go [hunk 20]
```diff
@@ -172,3 +174,72 @@ func (c *Client) DeleteSecurityGroup(ctx context.Context, id string) error {
...
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+// DeleteBucket deletes the s3 bucket with name <bucket>. If it does not exist,
+// no error is returned.
+func (c *Client) DeleteBucket(ctx context.Context, bucket string) error {
```
Comment (adracus) [inline controllers/provider-aws/pkg/aws/client/client.go]: I'd recommend renaming this to `DeleteBucketIfExists`

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+	return nil
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bb *extensionsv1alpha1.BackupBucket) error {
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(bb.Spec.SecretRef.Namespace, bb.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), bb.Spec.Region)
```
Comment (adracus) [inline controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: Instead of accessing the `providerSecret` directly, I'd recommend using `ReadCredentialsSecret`

File: controllers/provider-aws/pkg/controller/backupbucket/actuator.go (hunk 10 above)
Comment (adracus) [reply controllers/provider-aws/pkg/controller/backupbucket/actuator.go]: That's already there - `ReadCredentialsSecret`.

File: pkg/controller/backupentry/actuator.go (hunk 17 above)
Comment (swapnilgm) [reply pkg/controller/backupentry/actuator.go]: No. This is not erroneous case. Once shoot is deleted and backupEntry reconciliation is happening in deletionGracePeriod. Then seedNamespace won't be there and we don't want to deploy etcd backup Secret.

Also, at the start of shoot reconciliation, we deploy backupEntry resource independent of deploySeedNamesapce, So it might possible the seedNamespace is not created yet.

File: controllers/provider-aws/pkg/aws/client/client.go (hunk 19 above)
Comment (swapnilgm) [reply controllers/provider-aws/pkg/aws/client/client.go]: Agreed

File: pkg/controller/backupentry/mapper.go (hunk 18 above)
Comment (swapnilgm) [reply pkg/controller/backupentry/mapper.go]: The seedNamespace is not guaranteed to be there at the time of backupEntry reconciliation. So, it won't deploy etcd-backup secret if its not there. But later on when seedNamespace is created BackupEntry should be reconciled to deploy etcd-backup secret. So, i had to keep watch on namespace and map it to associated backupEntry.

**Comment by swapnilgm** [comment]
Thank you for the review and suggestion. I have addressed the comments. PTAL

File: controllers/provider-aws/pkg/aws/secret.go [hunk 22]
```diff
@@ -41,3 +46,19 @@ func ReadCredentialsSecret(secret *corev1.Secret) (*Credentials, error) {
 		SecretAccessKey: secretAccessKey,
 	}, nil
 }
+
+// NewClientFromSecretRef creates a new Client for the given AWS credentials from given k8s <secretRef> and
+// the AWS region <region>.
+func NewClientFromSecretRef(ctx context.Context, client client.Client, secretRef corev1.SecretReference, region string) (awsclient.Interface, error) {
```
Comment (rfranzke) [inline controllers/provider-aws/pkg/aws/secret.go]: Can you make https://github.com/gardener/gardener-extensions/blob/master/controllers/provider-aws/pkg/controller/infrastructure/actuator_delete.go#L59-L67 calling this?

File: pkg/controller/backupbucket/controller.go (hunk 12 above)
Comment (rfranzke) [reply pkg/controller/backupbucket/controller.go]: You haven't added it yet?!

File: pkg/controller/backupentry/genericactuator/genericactuator_suite_test.go [hunk 23]
```diff
@@ -0,0 +1,13 @@
+package genericactuator_test
```
Comment (rfranzke) [inline pkg/controller/backupentry/genericactuator/genericactuator_suite_test.go]: license header missing

File: pkg/controller/backupentry/genericactuator/types.go [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+
+const (
+	// EtcdBackupSecretName is the name of secret having credentials for etcd backups.
+	EtcdBackupSecretName string = "etcd-backup"
+
+	backupBucketName string = "bucketName"
+)
+
+// ProviderActuator preforms provider specific operation with BackupBucket resources.
+type ProviderActuator interface {
```
Comment (rfranzke) [inline pkg/controller/backupentry/genericactuator/types.go]: In the generic worker actuator we call this `WorkerDelegate`, do you want to rename it to `BackupEntryDelegate` ?

File: pkg/controller/backupentry/mapper.go (hunk 18 above)
Comment (rfranzke) [reply pkg/controller/backupentry/mapper.go]: Great, makes sense, thank you!

File: controllers/provider-aws/pkg/aws/secret.go (hunk 22 above)
Comment (swapnilgm) [reply controllers/provider-aws/pkg/aws/secret.go]: Ahh, I did but there anyways it require accessing provider secret for further terraform related calls. So, left as it is. May be we need to refactor terraform related code. I'll prefer having separate PR for that.

File: controllers/provider-aws/example/20-crd-backupbucket.yaml [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+  additionalPrinterColumns:
+  - name: Type
+    type: string
+    description: The type of the cloud provider for this resource.
+    JSONPath: .spec.type
+  - name: Region
+    type: string
+    description: The region into which the backup bucket should be created.
+    JSONPath: .spec.region
+  - name: Age
```
[the remark this answers is not in the document]
Comment (swapnilgm) [reply controllers/provider-aws/example/20-crd-backupbucket.yaml]: Sure

File: pkg/controller/backupentry/genericactuator/types.go (hunk 24 above)
Comment (swapnilgm) [reply pkg/controller/backupentry/genericactuator/types.go]: Sound good to me

File: pkg/controller/backupbucket/controller.go (hunk 12 above)
Comment (swapnilgm) [reply pkg/controller/backupbucket/controller.go]: I've added already. What else you want to add here?

File: controllers/provider-aws/pkg/aws/secret.go (hunk 22 above)
Comment (rfranzke) [reply controllers/provider-aws/pkg/aws/secret.go]: Ah, okay, got it. Makes sense, thanks.

File: pkg/controller/backupbucket/controller.go (hunk 12 above)
Comment (rfranzke) [reply pkg/controller/backupbucket/controller.go]: Hm, I refreshed the browser, now I see it. Strange, maybe cache issue in Chrome. Thanks!

File: pkg/controller/backupentry/genericactuator/actuator.go [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+		return err
+	}
+	etcdSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      EtcdBackupSecretName,
+			Namespace: shootTechnicalID,
+		},
+		Data: etcdSecretData,
+	}
+	return extensionscontroller.CreateOrUpdate(ctx, a.client, etcdSecret, nil)
```
Comment (rfranzke) [inline pkg/controller/backupentry/genericactuator/actuator.go]: Consistency with the way we use `CreateOrUpdate` elsewhere:

```suggestion
	return extensionscontroller.CreateOrUpdate(ctx, a.client, etcdSecret, func() error {
		etcdSecret.Data = etcdSecretData
	})
```

File: pkg/controller/backupentry/genericactuator/actuator.go [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+	etcdSecretData, err := a.backupEntryDelegate.GetETCDSecretData(ctx, be, backupSecretData)
+	if err != nil {
+		return err
+	}
+	etcdSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      EtcdBackupSecretName,
+			Namespace: shootTechnicalID,
+		},
+		Data: etcdSecretData,
```
Comment (rfranzke) [inline pkg/controller/backupentry/genericactuator/actuator.go]:
```suggestion
```

File: pkg/controller/backupentry/util.go [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+// limitations under the License.
+
+package backupentry
+
+import (
+	"strings"
+)
+
+// ExtractShootDetailsFromBackupEntryName returns Shoot resource technicalID its UID from provided <backupEntryName>.
+func ExtractShootDetailsFromBackupEntryName(backupEntryName string) (shootTechnicalID, shootUID string) {
```
Comment (rfranzke) [inline pkg/controller/backupentry/util.go]: Add a small test for this function, please :)

File: pkg/controller/backupentry/backupentry_suite_test.go [hunk 29]
```diff
@@ -0,0 +1,13 @@
+package backupentry_test
```
Comment (rfranzke) [inline pkg/controller/backupentry/backupentry_suite_test.go]: license header missing
